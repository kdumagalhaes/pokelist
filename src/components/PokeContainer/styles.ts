import styled from 'styled-components'

export const Container = styled.div`
  max-width: 54rem;
  margin: 3rem auto 15rem auto;
  display: flex;
  flex-direction: column;
  align-items: center;
`

export const Grid = styled.div`
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  padding: 0 1rem;
  gap: 2rem;
  cursor: pointer;
  margin-bottom: 6rem;

  @media only screen and (max-width: 1100px) {
    grid-template-columns: repeat(3, 1fr);
  }

  @media only screen and (max-width: 830px) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media only screen and (max-width: 600px) {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0 2rem;
  }
`
