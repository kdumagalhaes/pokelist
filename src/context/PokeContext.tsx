import {
  createContext,
  ReactNode,
  useContext,
  useState,
  useEffect,
  useCallback,
} from 'react'

interface Pokemon {
  id: number
  name: string
  url: string
}

interface PokeProviderProps {
  children: ReactNode
}

interface FavoritePokemon {
  id: number
  name: string
}
interface PokeContextModel {
  apiResponse: Pokemon[]
  getPokePicture: (id: string) => string
  nextPageUrl: string
  prevPageUrl: string
  goToNextPage: () => void
  goToPrevPage: () => void
  getPokeFavorites: (favoritePokemon: FavoritePokemon) => void
  favoritesList: FavoritePokemon[]
}

export const PokeContext = createContext({} as PokeContextModel)

export const PokeProvider = ({ children }: PokeProviderProps) => {
  const [apiResponse, setApiResponse] = useState<Pokemon[]>([])
  const [favoritesList, setFavoritesList] = useState<FavoritePokemon[]>([])
  const [currentPage, setCurrentPage] = useState(
    'https://pokeapi.co/api/v2/pokemon',
  )
  const [nextPageUrl, setNextPageUrl] = useState('')
  const [prevPageUrl, setPrevPageUrl] = useState('')

  // retrieve the basic pokemon data in the general list and pagination url
  useEffect(() => {
    fetch(currentPage)
      .then((res) => res.json())
      .then((json) => {
        setApiResponse(json.results)
        setNextPageUrl(json.next)
        setPrevPageUrl(json.previous)
      })
  }, [currentPage])

  useEffect(() => {
    const favoritesValue = localStorage.getItem('pokefavs')
    if (typeof favoritesValue === 'string') {
      const favorites = JSON.parse(favoritesValue)
      setFavoritesList(favorites)
    }
  }, [])

  const goToNextPage = () => {
    setCurrentPage(nextPageUrl)
  }

  const goToPrevPage = () => {
    setCurrentPage(prevPageUrl)
  }

  // retrieve id and name of the pokemon that the user has clicked on the fav button and save the data in localStore
  const getPokeFavorites = (favoritePokemon: FavoritePokemon) => {
    // if favoritesList doesn't have a pokemon with the same id as favoritePokemon, include it in the list
    if (!favoritesList.find((pokemon) => pokemon.id === favoritePokemon.id)) {
      localStorage.setItem(
        'pokefavs',
        JSON.stringify([...favoritesList, favoritePokemon]),
      )
    }
  }

  // retrieve the pokemon picture based on its id, which is presented in the general pokemon list
  const getPokePicture = useCallback((id: string) => {
    return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${id}.png`
  }, [])

  const value = {
    apiResponse,
    getPokePicture,
    nextPageUrl,
    prevPageUrl,
    goToNextPage,
    goToPrevPage,
    getPokeFavorites,
    favoritesList,
  }

  return <PokeContext.Provider value={value}>{children}</PokeContext.Provider>
}

export const usePokemon = () => {
  const context = useContext(PokeContext)

  if (context === undefined) {
    throw new Error('usePokemon must be used within PokeContext')
  }

  return context
}
