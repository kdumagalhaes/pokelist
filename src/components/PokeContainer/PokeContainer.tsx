import { usePokemon } from '../../context/PokeContext'
import { Pagination } from '../Pagination/Pagination'
import { PokeCard } from '../PokeCard/PokeCard'
import { Container, Grid } from './styles'

export function PokeContainer() {
  const { apiResponse, getPokePicture } = usePokemon()
  return (
    <Container>
      <Grid>
        {apiResponse.map((pokemon) => {
          return (
            <PokeCard
              key={pokemon.name}
              name={pokemon.name}
              image={getPokePicture(pokemon.url.slice(34, -1))}
            />
          )
        })}
      </Grid>
      <Pagination />
    </Container>
  )
}
