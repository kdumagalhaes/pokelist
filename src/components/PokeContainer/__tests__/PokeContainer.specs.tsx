import { render } from '@testing-library/react'
import { PokeContainer } from '../PokeContainer'
import { PokeProvider } from '../../../context/PokeContext'

describe('PokeContainer', () => {
  it('initial component should match snapshot', () => {
    const { asFragment } = render(
      <PokeProvider>
        <PokeContainer />
      </PokeProvider>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
