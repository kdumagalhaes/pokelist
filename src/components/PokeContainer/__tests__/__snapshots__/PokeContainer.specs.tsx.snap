// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeContainer initial component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dkrFOg hMYgDk"
  >
    <div
      class="sc-hLBbgP dDqwnA"
    />
    <div
      class="sc-bcXHqe cvRPSc"
    >
      <button
        class="pagination-btn"
      >
        <img
          alt=""
          class="pagination-icon"
          src="circle-left.svg"
        />
        Previous
      </button>
      <button
        class="pagination-btn"
      >
        <img
          alt=""
          class="pagination-icon"
          src="circle-right.svg"
        />
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
